src/LibraryManager.Contracts/FileHelpers.cs

```diff
@@ -17,6 +17,42 @@ namespace Microsoft.Web.LibraryManager.Contracts
     /// </summary>
     public static class FileHelpers
     {
+        /// <summary>
+        /// Writes the stream to a temporary file first, then moves the temporary file to the destination file
+        /// </summary>
+        /// <param name="fileName"></param>
+        /// <param name="sourceStream"></param>
+        /// <param name="cancellationToken"></param>
+        /// <returns></returns>
+        public static async Task<bool> SafeWriteToFileAsync(string fileName, Stream sourceStream, CancellationToken cancellationToken)
+        {
+            cancellationToken.ThrowIfCancellationRequested();
+
+            string tempFileName = Path.Combine(
+                Environment.GetFolderPath(Environment.SpecialFolder.InternetCache),
+                "libman",
+                fileName.GetHashCode().ToString(),
+                Path.GetFileName(fileName));
+
+            bool result = await WriteToFileAsync(tempFileName, sourceStream, cancellationToken).ConfigureAwait(false);
+
+            if (result)
+            {
+                result = await MoveFileAsync(tempFileName, fileName);
+            }
+
+            try
+            {
+                DeleteFiles(new string[] { tempFileName });
```

Comment:
>DeleteFiles [](start = 16, length = 11)

This should only be necessary when `!result`, since the tempFile would have been moved.  Is it worth putting in a conditional block?

Reply:
It actually also deletes empty folders. I wanted to clean up the "hash" folders we create for the file. So if the file got moved successfully, it'll delete empty folder. If not, it'll delete that temp file and folder.

Reply:
Made it conditional since I don't create a subfolder anymore